Refactor Dataset to use Compose for transforms

surajpaib opened the pull request:
Fixes #7646 

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

ericspod commented:
Hi @surajpaib thanks for the contribution. I think that the other classes inheriting from `Dataset` should be modified as well to pass along these arguments. They should have kwargs added to their constructors which is then passed to the superconstructor calls. There are other dataset classes which call `apply_transform` which maybe should be modified as well. The tests I think need to be a bit more thorough, currently they check the type of produced data which isn't indicative of the differing behaviour, you should be inspecting the return value somehow to demonstrate the added arguments have made a difference.

KumoLiu commented on monai/data/dataset.py:
```diff
@@ -77,15 +69,17 @@ class Dataset(_TorchDataset):
         self.data = data
-        self.transform: Any = transform
+        try:
+            self.transform = Compose(transform) if not isinstance(transform, Compose) else transform
```
Looks we miss support for transform to be None here.

surajpaib replied:
I've intentionally left this out as the `Compose` should handle the case where `transform` happens to be `None`. 

See:
https://github.com/Project-MONAI/MONAI/blob/e5afa4341b18149ca8d44f680f669b33ab423722/monai/transforms/compose.py#L238

https://github.com/Project-MONAI/MONAI/blob/e5afa4341b18149ca8d44f680f669b33ab423722/monai/transforms/compose.py#L105

KumoLiu replied:
Still suggest add a check in the dataset, it's not easy to understand even it's None, it will still return a compose object. What do you think?

surajpaib replied:
Is there any reason we need to have an explicit `None` check here?

Since `Compose` will handle the case when `transform = None` internally by assigning `self.transforms = []` which will give the exact same output as before (returning the data as is), I do not see the need to add an explicit `None` check. 

Moreover, the `Compose` logic is now consistent across dataset implementations making the behaviour much more understandable in my opinion.

KumoLiu replied:
Yes, I know the point, just feel it's not straightforward when users check ds.transform, it will still return a compose object.
```
import torch
from monai.data import Dataset

data = [torch.rand(10, 10) for _ in range(100)]
ds = Dataset(data=data, transform=None)
print(ds.transform) # --> <monai.transforms.compose.Compose object at 0x7ffa6088b250>
print(ds.transform.transforms) # --> ()
```
But it's ok if other developers also agree on this.
cc @ericspod @atbenmurray @Nic-Ma

ericspod replied:
I think this behaviour without a check is fine so long as the docstring line documenting `transform` states this, currently it makes no mention of what `None` would imply or what a sequence of callables would imply.

surajpaib replied:
@ericspod @KumoLiu Docstring for this is updated in the latest commit.

KumoLiu commented:
ci error log:
```
======================================================================
ERROR: test_shape_2 (tests.test_arraydataset.TestArrayDataset)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/parameterized/parameterized.py", line 620, in standalone_func
    return func(*(a + p.args), **p.kwargs, **kw)
  File "/Users/runner/work/MONAI/MONAI/tests/test_arraydataset.py", line 100, in test_shape
    data1 = dataset[0]
  File "/Users/runner/work/MONAI/MONAI/monai/data/dataset.py", line 1367, in __getitem__
    return self.dataset[index]
  File "/Users/runner/work/MONAI/MONAI/monai/data/dataset.py", line 106, in __getitem__
    return self._transform(index)
  File "/Users/runner/work/MONAI/MONAI/monai/data/dataset.py", line 1262, in _transform
    data.extend(to_list(dataset[index]))
  File "/Users/runner/work/MONAI/MONAI/monai/data/dataset.py", line 106, in __getitem__
    return self._transform(index)
  File "/Users/runner/work/MONAI/MONAI/monai/data/dataset.py", line 92, in _transform
    return self.transform(data_i)
TypeError: __call__() missing 1 required positional argument: 'lazy'

----------------------------------------------------------------------
```

surajpaib commented:
Hi @ericspod @KumoLiu 

The updated PR is ready for review.

Noting some of the changes from the PR below,
1. All dataset `transform`s use `Compose` now. I've replaced the `apply_transform` way of applying the transforms with `Compose` for the `Dataset`, `ZipDataset` and `NPZDictItemDataset` classes.
2. Removed the `Compose` checks from `PersistentDataset` and `CacheDataset` as these are now done in the `Dataset` base class.
3. Added tests for the `map_items=False` and `tuple` input scenario for `Dataset` in `test_dataset.py`. Modified tests for `test_profiling.py` and `test_arraydataset.py` for the `Compose` changes made. The functionality tested does not change. 

Overall, I think these changes make the `transform` application across `Datasets` more uniform and simplify underlying code as well.

surajpaib commented on monai/data/dataset.py:
```diff
@@ -323,9 +315,6 @@ def _pre_transform(self, item_transformed):
             random transform object
 
         """
-        if not isinstance(self.transform, Compose):
```
Since both these datasets inherit from `Dataset` and the check is now moved to `Dataset`, this point in the code would always receive a `Compose` object, making this check redundant.

KumoLiu replied:
When considering it along with the previous question, `PersistentDataset` must ensure that the transform is always not None. However, it will always convert to a compose in the base class, even with an empty transform, results in behavior that is inconsistent with the previous implementation.

ericspod replied:
Correct me if I'm wrong but I wonder if these checks were redundant in the original class as well? The lines in the constructor that this PR removes were converting `transform` to a `Compose` object if it wasn't already, so that is the behaviour that now is inherited. In these methods we're checking that `self.transform` is a `Compose` object, but wouldn't that always be the case in the existing code and in this PR? Am I missing some situation where this check would prevent some condition?

KumoLiu replied:
No, you are right. The previous implementation is also redundant. Mark as resolved.

KumoLiu commented:
/build

atbenmurray commented on monai/data/dataset.py:
```diff
@@ -77,15 +69,19 @@ class Dataset(_TorchDataset):
             data: input data to load and transform to generate dataset for model.
-            transform: a callable data transform on input data.
-
+            transform: a callable, sequence of callables or None. The transform will be passed
+            into a `Compose` wrapper by default. Sequence of callables are applied in order and
+            if `None` is passed, the data is returned as is.
```
```suggestion
            transform: a callable, sequence of callables or None. If transform is not
            a `Compose` instance, it will be wrapped in a `Compose` instance. Sequences
            of callables are applied in order and if `None` is passed, the data is returned as is.
```

ericspod reviewed (approved):
I'm good with this now if everyone else is.

KumoLiu commented:
/build
